stack-graphs/src/cycles.rs

```diff
@@ -127,3 +135,150 @@ where
         true
     }
 }
+
+// ----------------------------------------------------------------------------
+// Cycle detector
+
+pub trait Appendable {
+    type Ctx;
+
+    fn append_to(
+        &self,
+        graph: &StackGraph,
+        partials: &mut PartialPaths,
+        ctx: &mut Self::Ctx,
+        path: &mut PartialPath,
+    ) -> Result<(), PathResolutionError>;
+    fn start_node(&self, ctx: &mut Self::Ctx) -> Handle<Node>;
+    fn end_node(&self, ctx: &mut Self::Ctx) -> Handle<Node>;
+}
+
+#[derive(Clone)]
+pub struct AppendingCycleDetector<A> {
+    appendages: List<A>,
+}
+
+pub type Appendages<A> = ListArena<A>;
+
+impl<A: Appendable + Clone> AppendingCycleDetector<A> {
+    pub fn new() -> Self {
+        Self {
+            appendages: List::empty(),
+        }
+    }
+
+    pub fn from(appendages: &mut Appendages<A>, appendage: A) -> Self {
+        let mut result = Self::new();
+        result.appendages.push_front(appendages, appendage);
+        result
+    }
+
+    pub fn append(
+        &mut self,
+        graph: &StackGraph,
+        partials: &mut PartialPaths,
+        ctx: &mut A::Ctx,
+        appendages: &mut Appendages<A>,
+        appendage: A,
+    ) -> Result<(), PathResolutionError> {
+        let end_node = appendage.end_node(ctx);
+        self.appendages.push_front(appendages, appendage);
+
+        let mut maybe_cyclic_path = None;
+        let mut index = self.appendages;
+        let mut values = self.appendages;
+        loop {
+            // find loop point
+            let mut count = 0usize;
+            loop {
+                match index.pop_front(appendages) {
+                    Some(appendage) => {
+                        count += 1;
+                        if appendage.start_node(ctx) == end_node {
+                            break;
+                        }
+                    }
+                    None => return Ok(()),
+                }
+            }
+
+            // get prefix edges
+            let mut prefix_appendages = List::empty();
+            for _ in 0..count {
+                let appendage = values.pop_front(appendages).unwrap();
+                prefix_appendages.push_front(appendages, appendage.clone());
+            }
```

Comment:
Can you fold this into the above loop?  Build up `prefix_appendages` directly instead of `count`?

Also, I think the clone is unneeded, since you're not using `appendage` in the rest of the loop body.